src/lib/ingest.js

```diff
@@ -1,53 +1,264 @@
+const { addCollectionLinks, addItemLinks } = require('./api')
+const logger = console
+
+const COLLECTIONS_INDEX = process.env.COLLECTIONS_INDEX || 'collections'
+const H_LINK_KEYS = [
+  'self',
+  'root',
+  'parent',
+  'child',
+  'collection',
+  'item',
+  'items'
+]
+
+class ElasticSearchIngestClient {
+  constructor(config) {
+    this.config = config || {}
+    this.collections = []
+    if (!this.config.endpoint) {
+      logger.warn(
+        'No endpoint url defined, unable to replace URLs in successful items'
+      )
+    }
+  }
+
+  static async newClient(config) {
+    const client = new ElasticSearchIngestClient(config)
+    try {
+      await client.initClient()
+    } catch (err) {
+      logger.error(
+        `Failed to initalize ElasticSearchIngestClient: ${JSON.stringify(err)}`
+      )
+      throw err
+    }
+    return client
+  }
+
+  async initClient() {
+    this.client = await esClient.client()
+    await this.updateCollections()
+  }
+
+  async updateCollections() {
+    const indexResp = await this.client.cat.indices({
+      format: 'json',
+      h: 'index'
+    })
+
+    this.collections = new Set(indexResp.body.map((index) => index.index))
+  }
+
+  addLinks(originalRecord, convertedRecord) {
+    // convertedRecord has properties modified by itemToES,
+    // so we want to use it for the post-ingest handlers.
+    // However, it has the links stripped, so if the endpoint
+    // is not defined we need to get the links off the original
+    // because we can't use add links functions without the endpoint.
+    if (!this.config.endpoint) {
+      convertedRecord.links = originalRecord.links
+      return convertedRecord
+    }
+
+    let record = convertedRecord
+    if (convertedRecord.type === 'Feature') {
+      record = addItemLinks([convertedRecord], this.config.endpoint)[0]
+    } else if (convertedRecord.type === 'Collection') {
+      record = addCollectionLinks([convertedRecord], this.config.endpoint)[0]
+    } else {
+      logger.warn(`Cannot add links: unknown record type '${convertedRecord.type}'`)
+    }
+    return record
+  }
+
+  static get hLinkKeys() {
+    return H_LINK_KEYS
+  }
+
+  _toEs(item) {
+    // remove any hierarchy links in a non-mutating way
+    const hlinks = ElasticSearchIngestClient.hLinkKeys
+    const links = item.links.filter((link) => !hlinks.includes(link.rel))
+    const esDataObject = { ...item, links }
+
+    if (item.hasOwnProperty('properties')) {
+      const now = new Date().toISOString()
+      esDataObject.properties.created = now
+      esDataObject.properties.updated = now
+    }
+
+    return esDataObject
+  }
+
+  collectionToEs(collection) {
+    logger.debug(`Collection: ${JSON.stringify(collection)}`)
+    return {
+      doc: this._toEs(collection),
+      doc_as_upsert: true
+    }
+  }
+
+  itemToEs(item) {
+    logger.debug(`Item: ${JSON.stringify(item)}`)
+    const index = item.collection
+    if (!this.collections.has(index)) {
```

Comment:
Rather than trying to maintain this collections state, I think we'd be better off just inserting the item and catching an error if the index does not exist.

Reply:
That's fair, I was just trying to keep with the existing functionality.